docs/design/storage-auto-scale.md

```diff
@@ -0,0 +1,94 @@
+# Storage Automatic Scaler
+
+# Summary
+
+This feature would help users to get rid of maintenance task of keeping the system with enough available free space.
+In virtualized environments, we can scale the storage devices OSDs so this feature helps in automatic scaling. 
+
+# Proposal
+
+Implement auto storage scale CR per device class of a specific Storagecluster.
+
+# High-Level Design
+
+- Create a StorageAutoScale CR per Storagecluster device class
+    ```
+    apiVersion: ocs.openshift.io/v1
+    Kind: StorageAutoScaling
+    Metadata:
+        name: <storageclusterName>-<deviceClassName>
+        namespace: openshift-storage
+    Spec:
+        storageClusterName: <storageclusterName>
+        deviceClassName: <deviceClassName>
+        capacityLimit: <capacity>   
+    Status:
+        phase: (waiting/inProgress/completed/failure)
+        LastRunTimeStamp: (only on completed/failure)
+        currentOSDSize: 2TiB
+        expectedOSDSize: 4TiB
+        currentOSDCount: 3​
+        expectedOSDCount: 3
+    ```
+
+- Watch prometheus periodically after 10 mins and depending on the OSD fill percentage, reconcile the StorageAutoScaling CR.
+  
+- If the OSD size is less than 8Tib, do vertical scaling by doubling the each osd sizes for that device class.
+
+- If the OSD size is equal to 8Tib, do a horizontal scaling, by adding 1 osd of 8Tib on each storage node, by simply increasing the (count+1) in the device set associated with a device class.   
+
+# Low-Level Design 
+
+1) Add a single go-routine to watch periodically(10mins) all the OSDs sizes using Prometheus metrics in the storagecluster namespace.
```

Comment:
Alternative suggestion: don't use a goroutine at all. Instead, use the controller-runtime `requeueAfter` feature to re-run the reconcile loop every 10 minutes.

https://github.com/kubernetes-sigs/controller-runtime/blob/6f611112490242f9e8d5b0ccdc0f3bb8c86c85ef/pkg/reconcile/reconcile.go#L30-L37

Reply:
Good idea. With a new controller for the new CRD, and it could just constantly requeue every 10 minutes when not expanding, and every 1 minute to check status when expansion is in progress.

Reply:
That's a good idea,

But with this approach we need to query Prometheus with every CR,
And we cant use the same approach we were doing to query the prometheus once for all the cr's

Reply:
We should not do a `requestAfter` and run the entire reconcile just for a querying prom. It will basically make this controller into an infinite loop which is a bad practice. The only thing that needs iteration is the querying of prometheus and  my point of view is that we should isolate this as its own thing that runs in a loop and awakens the reconciler whenever it detect a change

Reply:
We are going to reconcile in an infinite loop with 10 minute retry period regardless of which implementation is used. I don't see why we should pick the more complicated implementation.

Reply:
Let's not complicate the design, as we are achieving the same goal from both perspectives.